src/history/HistoryManagerImpl.cpp

```diff
@@ -97,7 +102,13 @@ writeCheckpointFile(Application& app, HistoryArchiveState const& has,
         app.getHistoryManager().isLastLedgerInCheckpoint(has.currentLedger));
     auto filename = publishQueueFileName(has.currentLedger);
     auto tmpOut = app.getHistoryManager().getTmpDir() / filename;
-    has.save(tmpOut.string());
+    {
+        OutputFileStream out(app.getClock().getIOContext(),
```

Comment:
I think this might break `dropSQLBasedPublish`. That's the only function where `finalize == true`. It looks like we've changed from writing a JSON file to a binary file. This is fine for the temporary dirty files, but I think if `finalize == true`, this temp binary file will become the "canonical JSON file", but not have a JSON type. It might also help to clarify more with comments and naming that the temporary internal checkpoint file has a different type than the external JSON file we publish. I think operationally this makes sense, we just need to be careful about it.

Reply:
this is a good point to double check, but I think we're okay here. "Finalized" checkpoints essentially represent a queue of ledgers to publish (previously we did this with a sql table), and can be in any format. Later down the road we load these files into HAS objects and use them to create actual snapshots that are going to be published (https://github.com/stellar/stellar-core/pull/4555/files#diff-55b0f3e8b5f28727e3fd43b6af8569da025006279be1082bfcf0aeec44d20e08R522).